### ISSUE TITLE
Repeated errors when mapping across large dataset using 

### ISSUE THREAD

**Issue by seawolf42** [potatolondon/djangae#921]
I am running a map function across a dataset with about 11M entries:

    djangae.contrib.mappers.defer_iteration(MyModel.objects.all(), save, _queue='iteration')

I am getting the following error many times in the logs, probably about once per 300 tasks created by the overall job:

```Traceback (most recent call last):
  File "/base/data/home/apps/s~myproject/2-7d.401283859385756011/sitepackages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/base/data/home/apps/s~myproject/2-7d.401283859385756011/sitepackages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/base/data/home/apps/s~myproject/2-7d.401283859385756011/sitepackages/djangae/views.py", line 67, in deferred
    run(request.body)
  File "/base/data/home/runtimes/python27_experiment/python27_lib/versions/1/google/appengine/ext/deferred/deferred.py", line 155, in run
    return func(*args, **kwds)
  File "/base/data/home/apps/s~myproject/2-7d.401283859385756011/sitepackages/djangae/contrib/mappers/defer.py", line 38, in _shard
    _queue=queue
  File "/base/data/home/runtimes/python27_experiment/python27_lib/versions/1/google/appengine/ext/deferred/deferred.py", line 277, in defer
    return task.add(queue, transactional=transactional)
  File "/base/data/home/runtimes/python27_experiment/python27_lib/versions/1/google/appengine/api/taskqueue/taskqueue.py", line 1213, in add
    return self.add_async(queue_name, transactional).get_result()
  File "/base/data/home/runtimes/python27_experiment/python27_lib/versions/1/google/appengine/api/apiproxy_stub_map.py", line 613, in get_result
    return self.__get_result_hook(self)
  File "/base/data/home/runtimes/python27_experiment/python27_lib/versions/1/google/appengine/api/taskqueue/taskqueue.py", line 2138, in ResultHook
    rpc.check_success()
  File "/base/data/home/runtimes/python27_experiment/python27_lib/versions/1/google/appengine/api/apiproxy_stub_map.py", line 579, in check_success
    self.__rpc.CheckSuccess()
  File "/base/data/home/runtimes/python27_experiment/python27_lib/versions/1/google/appengine/api/apiproxy_rpc.py", line 134, in CheckSuccess
    raise self.exception
CancelledError: The API call taskqueue.BulkAdd() was explicitly cancelled.
```

My best guess is that there is no time limit in the function that is adding groups of entities to the queue and that it is timing out at 10 minutes.

**Comment by davidwtbuxton** [comment]
Hi,

What version of Djangae are you using?

Is the error causing the defer iteration to stop? Or is it that this message is logged, but the request itself continues and succeeds?

Are there any other relevant messages in the logs? Looks like recent versions of defer_iteration [log a message when they are about to hit the maximum shard count][1].

Does lowering [the value for `max_shards_to_defer_in_this_task`][2] make any difference?

Dave

[1]: https://github.com/potatolondon/djangae/blob/master/djangae/contrib/mappers/defer.py#L54
[2]: https://github.com/potatolondon/djangae/blob/master/djangae/contrib/mappers/defer.py#L34

**Comment by seawolf42** [comment]
djangae==0.9.8

It *appears* to still be iterating, the job has been running for about 2 days now with the appearance of progress (my task queue is set to max 1 concurrent task while I experiment to keep costs from spiraling). Every time it hits a task that is queueing up more shards it hits this error. The only thing I'm not sure of is whether any items are being saved multiple times (ie the sharder is repeating subsets of the data) or whether real progress is being made on items that have not yet been iterated.

And forgive me if I'm mixing up terms, I understand conceptually what's going on but this is my first bit M/R job in GAE so I'm still figuring out nuances and lingo :)

**Comment by davidwtbuxton** [comment]
Well the defer_iteration function stuff has changed since 0.9.8 - it would be a good idea to try the very latest version which changes the logic slightly so that it doesn't wait for DeadlineExceededError to trigger the next batch of deferreds. If the error message you are seeing is a result of the final deferred failing after DeadlineExceededError has been raised, then the latest version of Djangae may be the fix.

Hmmm… looks like the changes were included for 0.9.9, so that version might be sufficient (rather than the head of the master branch).

**Comment by nealedj** [comment]
If updating djangae doesn't help, read on.

It might be that the call to get the keys is consistently taking too long when the offset gets too high.

We had this on a project and ended up altering the `_shard` function to use a greater than PK query instead of an offset. This means that we always have to order by primary key and track the last primary key rather than the offset.

`ids = list(keys_queryset.filter(pk__gt=last_pk)[:shard_size])`

Full sample:

```
def _shard(model, query, callback, shard_size, queue, last_pk=0):
    logger.debug("Sharding PKs for model %s into tasks on queue %s", model.__name__, queue)
    keys_queryset = model.objects.all()
    keys_queryset.query = query

    # make sure we add the ordering after the original query is applied
    keys_queryset = keys_queryset.order_by('pk').values_list('pk', flat=True)

    shards_deferred = 0
    max_shards_to_defer_in_this_task = 250  # number which we think we can safely do in 10 minutes

    while True:
        try:
            ids = list(keys_queryset.filter(pk__gt=last_pk)[:shard_size])
            logger.info("Ids in this shard are {}".format(" ,".join(ids)))

            # with no ids left to shard off, exit
            if not ids:
                logger.debug(
                    "Finished sharding PKs for model %s into tasks on queue %s.",
                    model.__name__, queue
                )
                return

            # Fire off the first shard
            defer(_process_shard, model, ids, callback, _queue=queue)
            shards_deferred += 1

            if shards_deferred >= max_shards_to_defer_in_this_task:
                logger.debug("Redeferring. Last PK was: %s", last_pk)
                raise Redefer()
            else:
                last_pk = ids[-1]
                logger.info("Updating last PK to {}".format(last_pk))

        except (DeadlineExceededError, Redefer):
            # If we run out of time, or have done enough shards that we might be running out of
            # time, then defer this function again, continuing from the last PK.
            defer(
                _shard,
                model,
                query,
                callback,
                shard_size,
                queue,
                last_pk=last_pk,
                _queue=queue
            )
            return
```

**Comment by seawolf42** [comment]
@davidwtbuxton will updating `djangae` on a running job throw things off? I.e. if I upload a new version that's using 0.9.9 and make it the default version so that all traffic is going to that one, will the existing job in progress be negatively impacted in any way? I'm about 4 days into what looks like it's going to be a 28 day job (because I'm throttling requests to keep it under control), and I'd rather not restart if I can avoid it.

**Comment by seawolf42** [comment]
@nealedj two questions about your code:

1. Isn't `keys_queryset.filter(pk__gt=last_pk)[:shard_size]` an anti-pattern, as the whole queryset has to be retrieved from the Datastore in order to slice the desired subset?
1. How is this different from just passing `shard_size` to the call to `djangae.contrib.mappers.defer_iteration`?

Thanks for the help!

**Comment by nealedj** [comment]
> Isn't keys_queryset.filter(pk__gt=last_pk)[:shard_size] an anti-pattern, as the whole queryset has to be retrieved from the Datastore in order to slice the desired subset?

No, the slice sets a limit on the query. It doesn't run the full query and then slice it in memory.

> How is this different from just passing shard_size to the call to djangae.contrib.mappers.defer_iteration?

The existing way makes a query using offset/count. The query performance scales linearly with the sum off offset + count. (https://cloud.google.com/appengine/docs/standard/python/datastore/gqlqueryclass). Using a greater than query eliminates the offset and should result in this query not getting slower at increased offsets.

**Comment by seawolf42** [comment]
That's great to know. Is that true for all queries, or are there exceptions?

In the past I had to change some queries because I was trying to do just that and as the size of the result set grew they stopped performing. Also, the second bullet in the opening section of [Easy Efficient Pagination for the Datastore](https://djangae.readthedocs.io/en/latest/pagination/) states that skipping results is slow... does that refer specifically (and only) to the `offset` in `[offset:offset+count]` slices? Are `[:count]` slices fine in general (assuming `count` is not overly large)?

Thanks for all your help, this is clearing up several loose ends in my understanding for `djangae`!

**Comment by nealedj** [comment]
> does that refer specifically (and only) to the offset in [offset:offset+count] slices?
Yes, offset typically means the same as skip in this context.

> Are [:count] slices fine in general (assuming count is not overly large)?
Yes, all this does is limit the number of results that will be returned by the query. So smaller counts would typically be more performant.

**Comment by Kazade** [comment]
@seawolf42 In regards to your question about updating Djangae. You could, but it won't have any effect unless you are still seeing errors in the logs where it's still trying to defer. If so I don't *think* it'll affect the running job, but I don't *know*.

**Comment by seawolf42** [comment]
@nealedj and @Kazade, thanks for all that, very helpful.

**Comment by Kazade** [comment]
I believe this is all resolved in 0.9.10 so I'm closing. Reopen if that's not the case :)
